v2/cmd/csv.go

```diff
@@ -0,0 +1,281 @@
+	goModules = append(mainModule, goModules...)
+	klog.InfoS("Done: found dependencies", "count", len(goModules))
+	if klog.V(3).Enabled() {
+		for _, goModule := range goModules {
+			klog.InfoS("dependency", "module", goModule.ImportPath, "version", goModule.Version, "srcDir", goModule.SrcDir)
+		}
+	}
+	f, err := os.Create(config.Module.Csv.Path)
+	if err != nil {
+		return errors.Wrapf(err, "Creating license csv file")
+	}
+	defer func() {
+		closeErr := f.Close()
+		if err == nil {
+			// When there are no other errors, surface close file error.
+			// Otherwise file content may not be flushed to disk successfully.
+			err = closeErr
+		}
+	}()
+	_, err = f.WriteString("# Generated by https://github.com/google/go-licenses/v2. DO NOT EDIT.\n")
+	if err != nil {
+		return err
+	}
+	licenseCount := 0
+	errorCount := 0
+	for _, goModule := range goModules {
+		report := func(err error, args ...interface{}) {
+			errorCount = errorCount + 1
+			errorArgs := []interface{}{"module", goModule.ImportPath}
+			errorArgs = append(errorArgs, args...)
+			klog.ErrorS(err, "Failed", errorArgs...)
+		}
+		var override configmodule.ModuleOverride
+		for _, o := range config.Module.Overrides {
+			if o.Name == goModule.ImportPath {
+				override = o
+			}
+		}
+		// When override.Version == "", the override apply to any version.
+		if override.Version != "" && override.Version != goModule.Version {
+			report(fmt.Errorf("override version mismatch: version %q!=%q", goModule.Version, override.Version))
+			continue
+		}
+		if override.Skip {
+			klog.InfoS("Skipped", "module", goModule.ImportPath)
+			continue
+		}
+		repo, errGetGithubRepo := goutils.GetGithubRepo(goModule.ImportPath)
+		// this is not immediately an error, because we might specify override.License.Url below
+		type licenseInfo struct {
+			spdxId        string // required
+			licensePath   string // optional, required when url is not supplied
+			url           string // optional
+			subModulePath string // optional
+			lineStart     int    // optional
+			lineEnd       int    // optional
+		}
+		hasReportedGetGithubRepoErr := false
+		writeLicenseInfo := func(info licenseInfo) error {
+			if len(info.spdxId) == 0 {
+				return fmt.Errorf("failed writeLicenseInfo: info.spdxId required")
+			}
+			url := info.url
+			if len(url) == 0 {
+				if len(info.licensePath) == 0 {
+					return fmt.Errorf("failed writeLicenseInfo: info.licensePath required when info.url is empty")
+				}
+				if repo == nil && !hasReportedGetGithubRepoErr {
+					// now we need to use repo, so this becomes a fatal error
+					report(errGetGithubRepo)
+					hasReportedGetGithubRepoErr = true // only report once
+					// when repo == nil, repo.RemoteUrl has fallback behavior to use local path,
+					// so keep running to show more information to debug.
+				}
+				licensePath := info.licensePath
+				if info.subModulePath != "" {
+					licensePath = info.subModulePath + "/" + info.licensePath
```

Comment:
P2

Good point, I am aware of this problem.

I am leaning towards making it windows compatible in a batch when someone actually requests for windows support, because this is most likely not the only problem.

One problem for file path join is that, we sometimes use the path for URI path and sometimes for file path. In windows, we have to cautiously separate them, because they are different.

Anyway, I should call out in readme this is not currently windows compatible